aeppl/scan.py

```diff
@@ -0,0 +1,332 @@
+from copy import copy
+from typing import Callable, Dict, List, Optional, Tuple
+
+import aesara
+import aesara.tensor as at
+import numpy as np
+from aesara.graph.basic import Apply, Variable, clone_replace
+from aesara.graph.opt import local_optimizer, out2in
+from aesara.scan.op import Scan
+from aesara.scan.utils import ScanArgs
+from aesara.tensor.var import TensorVariable
+
+from aeppl.abstract import MeasurableVariable
+from aeppl.logprob import _logprob
+from aeppl.opt import logprob_rewrites_db, rv_sinking_db
+
+
+class MeasurableScan(Scan):
+    """A placeholder used to specify a log-likelihood for a scan sub-graph."""
+
+    @classmethod
+    def create_node(cls, node, rv_outs):
+        mscan_op = copy(node.op)
+        mscan_op.__class__ = cls
+        new_outputs = [o.clone() for o in node.outputs]
+        new_node = Apply(mscan_op, node.inputs, new_outputs)
+        mscan_op.scan_args = ScanArgs.from_node(new_node)
+        return new_node
+
+
+MeasurableVariable.register(MeasurableScan)
+
+
+@local_optimizer([Scan])
+def push_out_rvs_from_scan(fgraph, node):
+    r"""Push `MeasurableVariable`\s out of `Scan` nodes.
+
+    When `MeasurableVariable`\s are created within the inner-graph of a `Scan` and
+    are not output to the outer-graph, we "push" them out of the inner-graph.
+    This helps us produce an outer-graph in which all the relevant
+    `MeasurableVariable`\s are accessible (e.g. for constructing a log-likelihood
+    graph).
+    """
+
+    if not isinstance(node.op, Scan):
+        return None
+
+    if isinstance(node.op, MeasurableScan):
+        return
+
+    curr_scanargs = ScanArgs(
+        node.inputs, node.outputs, node.op.inputs, node.op.outputs, node.op.info
+    )
+
+    # Find the un-output `MeasurableVariable`s created in the inner-graph
+    clients: Dict[Variable, List[Variable]] = {}
+
+    local_fgraph_topo = aesara.graph.basic.io_toposort(
+        curr_scanargs.inner_inputs, curr_scanargs.inner_outputs, clients=clients
+    )
+    unpushed_inner_rvs = []
+    for n in local_fgraph_topo:
+        if isinstance(n.op, MeasurableVariable):
+            unpushed_inner_rvs.extend(
+                [c for c in clients[n] if c not in curr_scanargs.inner_outputs]
+            )
+
+    if len(unpushed_inner_rvs) != 0:
+        # This means that user used function like
+        # def scan_fn(*args):
+        #     S_t = at.random.categorical()
+        #     Y_t = at.random.normal(arg1[S_t])
+        # return Y_t
```

Comment:
If we dis-allow this kind of functions we wouldn't need the push-out functionality anymore. Right ? Cause all of the outputs are 'required' to be pushed out by the user himself. This function will just act as a check in that case.

Reply:
The original "push-out" functionality would automatically add `S_t` as an output to a new, reformulate `Scan`, as I recall; however, in our case, that's not particularly useful, because we need value variables for all the `MeasurableVariable`s in the model, but there isn't a (good) way to specify a value variable for an inner-graph variable like `S_t`.

In other words, yes, I think you have the right idea.

The only way I could see this being done without somehow specifying a value variable for terms like `S_t` is if we never even required value variables.  Instead, these `logprob`-related functions would also return a `dict` mapping `MeasurableVariable`s to value variables that are automatically generated for them by the function.

Reply:
I think that's how `symbolic-pymc` originally used to calculate the log-probability of Scan ops, no?

https://github.com/pymc-devs/symbolic-pymc/blob/84e8d612c714f502f8d188c1766498f4ff7beecf/symbolic_pymc/theano/pymc3.py#L177

Reply:
> I think that's how `symbolic-pymc` originally used to calculate the log-probability of Scan ops, no?

Yeah, as I recall.